v2/yarpchttp/handler.go

```diff
@@ -52,193 +49,197 @@ type handler struct {
 	logger              *zap.Logger
 }
 
-func (h handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+func (h handler) ServeHTTP(w http.ResponseWriter, httpReq *http.Request) {
 	responseWriter := newResponseWriter(w)
-	service := popHeader(req.Header, ServiceHeader)
-	procedure := popHeader(req.Header, ProcedureHeader)
-	bothResponseError := popHeader(req.Header, AcceptsBothResponseErrorHeader) == AcceptTrue
+	service := popHeader(httpReq.Header, ServiceHeader)
+	procedure := popHeader(httpReq.Header, ProcedureHeader)
+	bothResponseError := popHeader(httpReq.Header, AcceptsBothResponseErrorHeader) == AcceptTrue
 	// add response header to echo accepted rpc-service
-	responseWriter.AddSystemHeader(ServiceHeader, service)
-	err := h.callHandler(responseWriter, req, service, procedure)
-	status := yarpcerror.FromError(yarpcerror.WrapHandlerError(err, service, procedure))
-	if status == nil {
+	responseWriter.WriteSystemHeader(ServiceHeader, service)
+
+	res, resBuf, err := h.callHandler(responseWriter, httpReq, service, procedure)
+	if err == nil {
+		responseWriter.SetResponse(res, resBuf)
 		responseWriter.Close(http.StatusOK)
 		return
 	}
+
+	status := yarpcerror.FromError(yarpcerror.WrapHandlerError(err, service, procedure))
+	responseWriter.SetApplicationError()
+
 	if statusCodeText, marshalErr := status.Code().MarshalText(); marshalErr != nil {
 		status = yarpcerror.Newf(yarpcerror.CodeInternal, "error %s had code %v which is unknown", status.Error(), status.Code())
-		responseWriter.AddSystemHeader(ErrorCodeHeader, "internal")
+		responseWriter.WriteSystemHeader(ErrorCodeHeader, "internal")
 	} else {
-		responseWriter.AddSystemHeader(ErrorCodeHeader, string(statusCodeText))
+		responseWriter.WriteSystemHeader(ErrorCodeHeader, string(statusCodeText))
 	}
 	if status.Name() != "" {
-		responseWriter.AddSystemHeader(ErrorNameHeader, status.Name())
+		responseWriter.WriteSystemHeader(ErrorNameHeader, status.Name())
 	}
+
 	if bothResponseError && !h.legacyResponseError {
-		responseWriter.AddSystemHeader(BothResponseErrorHeader, AcceptTrue)
-		responseWriter.AddSystemHeader(ErrorMessageHeader, status.Message())
+		// Write the error message as a header AND set the response body. This is
+		// intended for returning structured errors (eg via proto.Any) and still
+		// getting the server error string.
+		responseWriter.WriteSystemHeader(BothResponseErrorHeader, AcceptTrue)
+		responseWriter.WriteSystemHeader(ErrorMessageHeader, status.Message())
+		responseWriter.SetResponse(res, resBuf)
 	} else {
-		responseWriter.ResetBuffer()
-		_, _ = fmt.Fprintln(responseWriter, status.Message())
-		responseWriter.AddSystemHeader("Content-Type", "text/plain; charset=utf8")
+		// Set the error message as the response body (not in a header)
+		responseWriter.WriteSystemHeader("Content-Type", "text/plain; charset=utf8")
```

Comment:
For my own understanding: do we not need the `Content-Type` in the previous case, even though we are writing to the response?

Reply:
We already set the content type at the end of `callHandlerI(..)`, so this step is actually stomping on that. In this specific case, we are making the error message the response body (aka not a structured response) so we need to indicate that it's `"text/plain"` and not `json`/`thrift` etc.